### ISSUE TITLE
Feature Request (for ReverseProxy): Add a forwarder header to include the path that was removed form url.

### ISSUE THREAD

**Issue by wasabii** [Azure/service-fabric-aspnetcore#38]
`UseReverseProxyIntegration` strips out the `/AppName/SvcName` components of the path just fine. But, it doesn't make them available from that point on. So, when further steps try to compose a URL, they end up directing to the wrong place.

Adding in `UseUniqueServiceUrl` will then add a new PathBase: with `/applicationId/partitionOrReplicaId`.

But there's no combination that seems to preserve `/AppName/SvcName`, for both reverse proxy and unique service URLs for stateless single instance services.

This works fine if you intend to ultimately host your reverse proxy services behind a further load balancer with a single root URL, but it doesn't work if you intend to direct users directly to the reverse proxy. Apps end up composing URLs that ultimately return 404s.

I would recommend some option which is LIKE `UseUniqueServiceUrl`, but instead preserves the `/AppName/SvcName` component of PathBase. Or, alternatively, to make the ReverseProxyIntegrationMiddleware detect what the path SHOULD BE given the type of service.

**Comment by amanbha** [comment]
UseReverseProxyIntegration doesn't strip out /AppName/SvcName, it would be done by the ReverseProxy so by the time call comes to these middlewares, the path will not have it. All this ReverseProxy middleware does is adds a specific header to response for 4o4 from the service.
example:
lets say a service with name fabric:/MyApp/MyService is deployed, it has a route called api/values
With UseUniqueServiceUrl, url for the endoiint reported to SF runtime is (
http://machinename:port/partitionId/repilcaId/guid

To call it directly, client will resolve the service partition and send the request as
http://machinename:port/partitionId/repilcaId/guid/api/values which would come to the ServiceFabricMiddleware and after adjusting the path and pathbase, call will end up at /api/values route.

To call it with ReverseProxy, call will be made as:
http://:Port/MyApp/MyService/api/values?PartitionKey=&PartitionKind=
ReverseProxy (not the ReverseProxyMiddleware) will resolve the service for the specified partition and send the request to
http://machinename:port/partitionId/repilcaId/guid/api/values which would come to the ServiceFabricMiddleware and after adjusting the path and pathbase, call will end up at /api/values route

**Comment by wasabii** [comment]
Okay, I understand that. The problem is I need to figure out how to get this information into the AspNetCore context, so the app can make decisions on how to construct absolute URLs into the app.

And in my particular case, I have clients which come through Reverse Proxy. So I need to generate URLs that use that.

Maybe I'm looking for another middleware set that figures THAT out?

**Comment by amanbha** [comment]
Sorey, the requirement here is not clear to me, could you give a more concrete example. 
From my understanding of it:
If you have clients coming through reverse proxy, then they must know the name of service, they intend to talk to and will provide it in URL to reverse proxy. If the clients are within the cluster, they don't need to go through reverse proxy, they can directly resolve the service and send request to it but they would still need service name they intend to talk to.

Service code can also get its service name by accessing ServiceContext

**Comment by wasabii** [comment]
The clients come through the reverse proxy. In the form of http://cluster:19081/AppName/ServiceName/relative/path

But, the application needs to be able to generate full URLs. It might output HTML that needs a full URL. Or, it might need to generate a Redirect to another URL.

So, ASP.Net needs to be aware of the full URL.

Yes, it could access the ServiceContext directly. But ASP.Net core code shouldn't have access to that. Instead, it should be using stuff like:

```
public IActionResult GetThing()
{
  return Redirect(Request.PathBase + "/other/url");
}
```

So, what I'm saying, is that in order to shield the ASP.net code from working with ServiceContext directly, or being aware of any of that, something ELSE, some middleware, should be setting up PathBase to contain the right value.

**Comment by amanbha** [comment]
@wasabii  When the request comes to the service, the information is not available in the path, it has been stripped out by reverseproxy, so middleware cannot put that info as it has no idea what it was. The service doesn't know whether the request came from ReverseProxy or from some other client directly and it should not depend on that.

**Comment by wasabii** [comment]
Right, but ASP.Net does regularly depend on that. And that's just the way it is.

When you're outputting HTML, you quite frequently want to output working absolute paths, relative to the client's request. When you're redirecting, you need to know the proper path to redirect to. And the established convention is the PathBase should contain that info.

So I'm trying to figure out how to get it there.

**Comment by amanbha** [comment]
The requests can come from either SF reverse proxy, or directly or through some other custom proxy. So the middleware cannot figure out generically if its stripped out of the request. May be it can be put in some well known header by SF reverse proxy and then you write a middleware to use it the way your service wants OR your middleware can get it from service context.
Adding @vturecek for scenario and @kavyako for SF reverse proxy.

**Comment by wasabii** [comment]
One convention is that the reverse proxy or load balancer sets a number of well-known headers, like `X-Forwarded-For`, `X-Forwarded-Host`, `X-Forwarded-Scheme`. And then AspNet middleware checks those headers, and replaces HttpRequest.Host, and the various variables so that the service "thinks", for instance, that the host name is the original name. And not the internal name the load balancer might be referring to it as. There's existing standard AspNet middleware for these headers specifically.

As it goes for the base path, there's a handful of conventions like X-Original-URL, and then additional middleware that will find the PathBase from the difference between that and the actual path being invoked.

It looks like Service Fabric reverse proxy is setting X-Forwarded- headers properly. But there's nothing being passed that's path-related. So, I feel like some SF-specific middleware should be able to detect whether the RP is being used and deal with that specifically.

I have no idea how to deal with the `PartitionId` query args if they're required. But I can ignore that problem for now because I have a single stateless service.

**Comment by wminos** [comment]
My temporary solution for local service fabric cluster. Maybe it's not nice.
I want to know best practice for service fabric. 
The service fabric is currently under review and i don't know that well.

```C#
            // rewrite path base if request is from reverse proxy like 19081 port (temporary)
            var rewriteOptions = new RewriteOptions()
               .Add(rewriteContext =>
               {
                   var request = rewriteContext.HttpContext.Request;
                   if (request.Headers.ContainsKey("X-Forwarded-Host"))
                   {
                       try
                       {
                           var fabricServiceName = Environment.GetEnvironmentVariable("Fabric_ServiceName"); // ex) fabric:/AppName/ServiceName
                           var fabricServiceUri = new Uri(fabricServiceName);
                           var servicePathBase = fabricServiceUri.AbsolutePath; // ex) /AppName/ServiceName
                           request.PathBase = servicePathBase;
                       }
                       catch (Exception)
                       {
                       }
                   }
               });
            app.UseRewriter(rewriteOptions);
```

**Comment by wasabii** [comment]
I ended up with something even a little worse than @wminos' solution. I actually detect the target port added into X-Forwarded-Host, and check whether it matches with the Reverse Proxy as configured within the cluster manifest.

This is more a bug for the Service Fabric Reverse Proxy: It should pass an additional header with the original path. Either using a convention (X-Forwarded-*), or a special header we could at least use middleware against.
